includes/reader-activation/class-esp-sync-admin.php

```diff
@@ -0,0 +1,216 @@
+		if ( self::ADMIN_ACTION !== $doaction ) {
+			return $sendback;
+		}
+		// Bulk action requires ActionScheduler, so we don't show it if it's not available.
+		if ( ! class_exists( 'ActionScheduler' ) ) {
+			\wp_die( \esc_html__( 'ActionScheduler is not available to perform the bulk action.', 'newspack-plugin' ) );
+		}
+		$can_sync = self::can_esp_sync( true );
+		if ( $can_sync->has_errors() ) {
+			\wp_die( $can_sync ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		}
+		if ( ! \current_user_can( 'edit_users' ) ) {
+			\wp_die( \esc_html__( 'You do not have permission to do that.', 'newspack-plugin' ) );
+		}
+		foreach ( $items as $user_id ) {
+			as_schedule_single_action( time(), 'newspack_sync_admin_batch', [ 'user_id' => $user_id ] );
+		}
+		$sendback = \add_query_arg(
+			[
+				'update'                  => self::ADMIN_ACTION,
+				'scheduled-sync-contacts' => count( $items ),
+			],
+			$sendback
+		);
+		return $sendback;
+	}
+
+	/**
+	 * Process admin action request.
+	 */
+	public static function process_admin_action() {
+
+		if ( ! isset( $_GET['action'] ) || self::ADMIN_ACTION !== $_GET['action'] ) {
+			return;
+		}
+
+		$action = \sanitize_text_field( \wp_unslash( $_GET['action'] ) );
+
+		// If we don't have UID, it's probably a bulk action.
+		if ( ! isset( $_GET['uid'] ) ) {
+			return;
+		}
+
+		if ( ! \check_admin_referer( $action ) ) {
+			\wp_die( \esc_html__( 'Invalid request.', 'newspack-plugin' ) );
+		}
+
+		$user_id = \absint( \wp_unslash( $_GET['uid'] ) );
+
+		if ( ! \current_user_can( 'edit_user', $user_id ) ) {
+			\wp_die( \esc_html__( 'You do not have permission to do that.', 'newspack-plugin' ) );
+		}
+
+		$user = \get_user_by( 'id', $user_id );
+
+		if ( ! $user || \is_wp_error( $user ) ) {
+			\wp_die( \esc_html__( 'User not found.', 'newspack-plugin' ) );
+		}
+
+		$result = self::sync_contact( $user_id );
+		if ( \is_wp_error( $result ) ) {
+			\wp_die( $result ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		}
+
+		$redirect = \add_query_arg( [ 'update' => $action ], \remove_query_arg( [ 'action', 'uid', '_wpnonce', 'scheduled-sync-contacts' ] ) );
+		\wp_safe_redirect( $redirect );
+		exit;
+	}
+
+	/**
+	 * Admin notices.
+	 */
+	public static function admin_notices() {
+		// phpcs:disable WordPress.Security.NonceVerification
+		if ( ! isset( $_GET['update'] ) ) {
+			return;
+		}
+		$update = sanitize_text_field( wp_unslash( $_GET['update'] ) );
+		if ( self::ADMIN_ACTION !== $update ) {
+			return;
+		}
+		$message = __( 'Contact resynced to the ESP.', 'newspack-plugin' );
+		if ( isset( $_GET['scheduled-sync-contacts'] ) ) {
+			$scheduled_syncs = absint( wp_unslash( $_GET['scheduled-sync-contacts'] ) );
+			$message         = sprintf(
+				// translators: %1$d: number of contacts resynced. %2$d: estimated time in seconds.
+				__( '%1$d contacts scheculed for resync to the ESP. This should take approximately %2$d seconds.', 'newspack-plugin' ),
```

Comment:
Looking good. Nice that we don't show the option if ActionScheduler is not present.

As a suggestion, I think we should just say that the contacts are going to be synced in background and not give an estimate.

There is only 10 contacts in a page, so the max amount we are going to show is 20 seconds
1. 20 seconds doesn't sound that long to justify all the hassle
2. If it really were only 20 seconds it would not timeout.

Each action is processed by AS in an independent request, so there is the overhead of making a request, etc. It takes a little more.... and how long it takes is a bit unpredictable...

idk, just feel that this info there doesn't help.

"They will be processed in a little while..." idk

Reply:
We can say it'll take "a couple of minutes". It's a good enough estimate.

Reply:
Updated in 0d554eb994eb23d7b5f5df1b5aec61a5f6cce717